packages/gatsby/src/utils/source-nodes.ts

```diff
@@ -104,14 +105,19 @@ export default async ({
     webhookBody: webhookBody || {},
     pluginName,
   })
-  await getDataStore().ready()
 
-  const state = store.getState()
-  const nodes = getNodes()
+  await getDataStore().ready()
 
-  warnForPluginsWithoutNodes(state, nodes)
+  // We only warn for plugins w/o nodes and delete stale nodes on the first sourcing.
+  if (initialSourcing) {
+    const state = store.getState()
+    const nodes = getNodes()
 
-  deleteStaleNodes(state, nodes)
+    warnForPluginsWithoutNodes(state, nodes)
 
-  store.dispatch(actions.apiFinished({ apiName: `sourceNodes` }))
+    deleteStaleNodes(state, nodes)
+    initialSourcing = false
+  } else {
+    return
+  }
```

Comment:
seems to be some unnecessary code?
```suggestion
  }
```